File: MATTHEW/leetcode/AIcoding/5.Qwen-3/go_solutions/777-swap-adjacent-in-lr-string.go
package main

func canTransform(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}
	i, j := 0, 0
	for i < len(s) && j < len(t) {
		for i < len(s) && s[i] == 'X' {
			i++
		}
		for j < len(t) && t[j] == 'X' {
			j++
		}
		if i < len(s) && j < len(t) {
			if s[i] != t[j] {
				return false
			}
			if s[i] == 'L' && i < j {
				return false
			}
			if s[i] == 'R' && i > j {
				return false
			}
			i++
			j++
		} else {
			break
		}
	}
	return true
}

File: leetcode/AIcoding/1.Gemini-2.5/c_solutions/777-swap-adjacent-in-lr-string.c
#include <string.h>
#include <stdbool.h>

bool canTransform(char* start, char* end) {
    int n = strlen(start);
    int m = strlen(end);

    if (n != m) {
        return false;
    }

    int i = 0; // pointer for start string
    int j = 0; // pointer for end string

    while (i < n && j < n) {
        // Skip 'X' characters in the start string
        while (i < n && start[i] == 'X') {
            i++;
        }
        // Skip 'X' characters in the end string
        while (j < n && end[j] == 'X') {
            j++;
        }

        // If both pointers reached the end, or one did while the other didn't (handled by final check)
        if (i == n || j == n) {
            break;
        }

        // If the non-'X' characters don't match, transformation is impossible
        if (start[i] != end[j]) {
            return false;
        }

        // Check movement constraints
        if (start[i] == 'L') {
            // 'L' can only move left. Its original position (i) must be greater than or equal to its final position (j).
            if (i < j) {
                return false;
            }
        } else { // start[i] == 'R'
            // 'R' can only move right. Its original position (i) must be less than or equal to its final position (j).
            if (i > j) {
                return false;
            }
        }

        // Move to the next non-'X' character pair
        i++;
        j++;
    }

    // After the loop, ensure both pointers have reached the end of their respective strings
    // (i.e., all characters, including trailing 'X's, have been processed).
    while (i < n && start[i] == 'X') {
        i++;
    }
    while (j < n && end[j] == 'X') {
        j++;
    }

    return i == n && j == n;
}